src/experiments/perf/api.js

```diff
@@ -0,0 +1,328 @@
+      "domainLookupEndTime",
+      "connectStartTime",
+      "tcpConnectEndTime",
+      "secureConnectionStartTime",
+      "connectEndTime",
+      "requestStartTime",
+      "responseStartTime",
+      "responseEndTime",
+      //"cacheReadStartTime",
+      //"cacheReadEndTime",
+      //"redirectStartTime",
+      //"redirectEndTime",
+    ].forEach((key) => {
+      result[key] = getFieldValue(channel, key);
+    });
+
+    channel.QueryInterface(Ci.nsIHttpChannelInternal);
+    result.remoteAddress = channel.remoteAddress;
+
+    result.status = getFieldValue(channel, "status");
+    //result.dnsLookupDiff = result.domainLookupEndTime - result.domainLookupStartTime;
+
+    let allHeaders = xhr.getAllResponseHeaders().split(/[\r\n]+/);
+
+    // If we are checking we got a cache hit
+    if (config.ccheck) {
+      let xcache = allHeaders.find((header) => {
+        return /^x-cache/i.test(header);
+      });
+      if (xcache && !/TCP_HIT|TCP_MEM_HIT/.test(xcache)) {
+        // Will retry request
+        return null;
+      }
+    }
+    result.headers = allHeaders.filter((header) => {
+      return /^synth[-]timings/i.test(header);
+    });
+
+    let securityInfo = getFieldValue(channel, "securityInfo");
+
+    if (securityInfo instanceof Ci.nsITransportSecurityInfo) {
+      securityInfo.QueryInterface(Ci.nsITransportSecurityInfo);
+
+      // extract security state and error code by which we can identify the reasons the connection failed
+      result.securityState = getFieldValue(securityInfo, "securityState");
+      result.errorCode = getFieldValue(securityInfo, "errorCode");
+    }
+  } catch (ex) {
+    result.exception = ex.message;
+  }
+
+  if (config.doh) {
+    let url = new URL(config.url);
+    try {
+      let res = await verifyWasTRR(url.hostname);
+      if (!res.isTRR) {
+        Cu.reportError("trr record not found");
+      }
+      result.isTRR = res.isTRR;
+    } catch {
+      Cu.reportError("dns didn't return");
+      return null;
+    }
+  }
+
+  return result;
+}
+
+
+function verifyWasTRR(hostname) {
+  return new Promise((resolve, reject) => {
+    let response = {};
+    let request;
+    let listener = {
+      onLookupComplete: function(inRequest, inRecord, inStatus) {
+        if (inRequest === request) {
+          if (!Components.isSuccessCode(inStatus)) {
+            return reject({message: getErrorString(inStatus)});
+          }
+          response.isTRR = inRecord.IsTRR();
+          return resolve(response);
+        }
+      },
+    };
+
+    let dns = Cc["@mozilla.org/network/dns-service;1"].getService(Ci.nsIDNSService);
+    try {
+      request = dns.asyncResolve(hostname, Ci.nsIDNSService.RESOLVE_OFFLINE, listener, null, {} /* defaultOriginAttributes */);
+    } catch (e) {
+      // handle exceptions such as offline mode.
+      return reject({message: e.name});
+    }
+  });
+}
+
+function buildRequest(config, reportResult) {
+  Services.prefs.setBoolPref(ECS_PREF, !!config.ecs);
```

Comment:
I presume this is getting cleaned up automatically by being included in the JSON file?

Reply:
Yeah anything in prefTypes is removed on addon uninstall.